Mathlib/Topology/Category/Profinite/Extend.lean

```diff
@@ -0,0 +1,207 @@
+its limit in `Profinite`. Let `G` be a functor from `Profinite` to a category `C` and suppose that
+`G` preserves the limit described above. Suppose further that the projection maps `S ⟶ Sᵢ` are
+epimorphic for all `i`. Then `G.obj S` is isomorphic to a limit indexed by
+`StructuredArrow S toProfinite` (see `Profinite.Extend.isLimitCone`).
+
+We also provide the dual result for a functor of the form `G : Profiniteᵒᵖ ⥤ C`.
+
+We apply this to define `Profinite.diagram'`, `Profinite.asLimitCone'`, and `Profinite.asLimit'`,
+analogues to their unprimed versions in `Mathlib.Topology.Category.Profinite.AsLimit`, in which the
+indexing category is `StructuredArrow S toProfinite` instead of `DiscreteQuotient S`.
+-/
+
+universe u w
+
+open CategoryTheory Limits FintypeCat Functor
+
+attribute [local instance] ConcreteCategory.instFunLike
+
+namespace Profinite
+
+variable {I : Type u} [Category.{u} I] [IsCofiltered I]
+    {F : I ⥤ FintypeCat.{max u w}} (c : Cone <| F ⋙ toProfinite)
+
+/--
+A continuous map from a profinite set to a finite set factors through one of the components of
+the profinite set when written as a cofiltered limit of finite sets.
+-/
+lemma exists_hom (hc : IsLimit c) {X : FintypeCat} (f : c.pt ⟶ toProfinite.obj X) :
+    ∃ (i : I) (g : F.obj i ⟶ X), f = c.π.app i ≫ toProfinite.map g := by
+  let _ : TopologicalSpace X := ⊥
+  have : DiscreteTopology (toProfinite.obj X) := ⟨rfl⟩
+  let f' : LocallyConstant c.pt (toProfinite.obj X) :=
+    ⟨f, (IsLocallyConstant.iff_continuous _).mpr f.continuous⟩
+  obtain ⟨i, g, h⟩ := exists_locallyConstant.{_, u} c hc f'
+  refine ⟨i, (g : _ → _), ?_⟩
+  ext x
+  exact LocallyConstant.congr_fun h x
+
+namespace Extend
+
+/--
+Given a cone in `Profinite`, consisting of finite sets and indexed by a cofiltered category,
+we obtain a functor from the indexing category to `StructuredArrow c.pt toProfinite`.
+-/
+@[simps]
+def functor : I ⥤ StructuredArrow c.pt toProfinite where
+  obj i := StructuredArrow.mk (c.π.app i)
+  map f := StructuredArrow.homMk (F.map f) (c.w f)
+
+-- We check that the original diagram factors through `Profinite.Extend.functor`.
+example : functor c ⋙ StructuredArrow.proj c.pt toProfinite ≅ F := Iso.refl _
+
+/--
+Given a cone in `Profinite`, consisting of finite sets and indexed by a cofiltered category,
+we obtain a functor from the opposite of the indexing category to
+`CostructuredArrow toProfinite.op ⟨c.pt⟩`.
+-/
+@[simps! obj map]
+def functorOp : Iᵒᵖ ⥤ CostructuredArrow toProfinite.op ⟨c.pt⟩ :=
+  (functor c).op ⋙ StructuredArrow.toCostructuredArrow _ _
+
+-- We check that the opposite of the original diagram factors through `Profinite.Extend.functorOp`.
+example : functorOp c ⋙ CostructuredArrow.proj toProfinite.op ⟨c.pt⟩ ≅ F.op := Iso.refl _
+
+/--
+If the projection maps in the cone are epimorphic and the cone is limiting, then
+`Profinite.Extend.functor` is initial.
+-/
+lemma functor_initial (hc : IsLimit c) [∀ i, Epi (c.π.app i)] : Initial (functor c) := by
```

Comment:
I think this is also true with the weaker assumption that for all i there exists a j mapping to i such that `c.pi.app j` is epi, right? The way you stated it is the right way to do it from the point of view of the typeclass system, but we may want to think later about how to formulate such weaker assumptions.

Reply:
I guess so. I'll add a TODO in the docstring